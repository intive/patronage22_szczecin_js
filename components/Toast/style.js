import styled, { css } from 'styled-components'
import Icon from '../Icon/Icon'
import { breakpoint } from '../../config/breakpoints'

export const NotificationToast = styled.div`
  position: fixed;
  z-index: 100;
  bottom: 2rem;
  right: 2rem;
  animation: fadein 0.5s, fadeout 0.5s 3.5s;
  display: flex;
  align-items: center;
  justify-content: left;
  box-shadow: 0px 1px 3px rgba(98, 98, 98, 0.24);
  border-radius: 4px;
  padding: 0.5rem 1rem 0.5rem 0.5rem;
  min-width: 11.25rem;
  max-width: 20rem;
  height: 2.5rem;
  background: #DC3545;
  font-size: 0.875rem;
  color: #FFFFFF;

  @keyframes fadein {
    from {bottom: 0rem; position: fixed; opacity: 0;}
    to {bottom: 2rem; position: fixed; opacity: 1;}
  }

  @keyframes fadeout {
    from {bottom: 2rem; position: fixed; opacity: 1;}
    to {bottom: 0rem; position: fixed; opacity: 0;}
  }

  @media only screen and ${breakpoint.md} {
    max-width: 100%;
  }

  ${props => props.type === 'success' && css`
    background: #28A745;
  `}

`

export const NotificationIcon = styled(Icon)`
  margin-right: 0.2rem;
`
