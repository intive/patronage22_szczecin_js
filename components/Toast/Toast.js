import { NotificationToast, NotificationIcon } from './style'
import ReactPortal from '../ReactPortal/ReactPortal'
import { ulid } from 'ulid'
import { useEffect, useState } from 'react'

const Toast = ({ id, type, children, isOpen }) => {
  const [portalId] = useState(ulid())
  const [isOpenToast, setIsOpenToast] = useState(isOpen)
  const autoCloseTime = 3900

  useEffect(() => {
    setTimeout(() => {
      setIsOpenToast(false)
    }, autoCloseTime)
  }, [])

  return isOpenToast
    ? (
      <ReactPortal wrapperId={`toast-portal-${id || portalId}`}>
        <NotificationToast type={type} role='alert' data-testid='toast' aria-hidden='true'>
          {type === 'success'
            ? <NotificationIcon name='done' />
            : <NotificationIcon name='priority_high' />}
          {children}
        </NotificationToast>
      </ReactPortal>
      )
    : null
}

export default Toast
