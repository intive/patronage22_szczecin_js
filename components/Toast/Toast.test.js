/* globals describe, expect, it */

import { act, render, screen } from '@testing-library/react'
import Toast from './Toast'
import { jest } from '@jest/globals'

describe('Toast', () => {
  it('should render Toast with given text', () => {
    const { baseElement } = render(
      <Toast isOpen id='toast-test-id' type='success'>Board added successfully</Toast>)

    expect(screen.getByText('Board added successfully')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render success Toast', () => {
    const { baseElement } = render(
      <Toast isOpen id='toast-test-id' type='success'>Board added successfully</Toast>)

    expect(screen.getByText(/done/, { consider: 'span' })).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render error Toast', () => {
    const { baseElement } = render(
      <Toast isOpen id='toast-test-id'>Failed to add board</Toast>)

    expect(screen.getByText(/priority_high/, { consider: 'span' })).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render Toast and automatically close it after 4 seconds', () => {
    jest.useFakeTimers()
    render(<Toast isOpen id='toast-test-id'>Failed to add board</Toast>)
    const toast = screen.getByText('Failed to add board')

    expect(toast).toBeInTheDocument()
    act(() => {
      jest.advanceTimersByTime(4000)
    })
    expect(toast).not.toBeInTheDocument()

    jest.clearAllTimers()
  })

  it('should render Toast, Toast should still be present after 3sec', () => {
    jest.useFakeTimers()
    render(<Toast isOpen id='toast-test-id'>Failed to add board</Toast>)
    const toast = screen.getByText('Failed to add board')

    expect(toast).toBeInTheDocument()
    act(() => {
      jest.advanceTimersByTime(3000)
    })
    expect(toast).toBeInTheDocument()

    jest.clearAllTimers()
  })
})
