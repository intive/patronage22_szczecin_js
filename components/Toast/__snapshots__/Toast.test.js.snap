// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render Toast with given text 1`] = `
<body>
  <div />
  <div
    id="toast-portal-toast-test-id"
  >
    <div
      aria-hidden="true"
      class="sc-bdvvtL kYuxJh"
      data-testid="toast"
      role="alert"
      type="success"
    >
      <span
        class="material-icons sc-gsDKAQ hHMBic"
      >
        done
      </span>
      Board added successfully
    </div>
  </div>
</body>
`;

exports[`Toast should render error Toast 1`] = `
<body>
  <div />
  <div
    id="toast-portal-toast-test-id"
  >
    <div
      aria-hidden="true"
      class="sc-bdvvtL jJTopw"
      data-testid="toast"
      role="alert"
    >
      <span
        class="material-icons sc-gsDKAQ hHMBic"
      >
        priority_high
      </span>
      Failed to add board
    </div>
  </div>
</body>
`;

exports[`Toast should render success Toast 1`] = `
<body>
  <div />
  <div
    id="toast-portal-toast-test-id"
  >
    <div
      aria-hidden="true"
      class="sc-bdvvtL kYuxJh"
      data-testid="toast"
      role="alert"
      type="success"
    >
      <span
        class="material-icons sc-gsDKAQ hHMBic"
      >
        done
      </span>
      Board added successfully
    </div>
  </div>
</body>
`;
